homeassistant/components/keenetic_ndms2/__init__.py

```diff
@@ -57,6 +61,24 @@ async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry) ->
 
     hass.data[DOMAIN].pop(config_entry.entry_id)
 
+    if router.tracked_interfaces - set(config_entry.options[CONF_INTERFACES]):
+        _LOGGER.debug(
+            "Removing device_tracker entities since some interfaces are now untracked"
+        )
+        ent_reg = await entity_registry.async_get_registry(hass)
+        dev_reg = await device_registry.async_get_registry(hass)
+        for entity_entry in list(ent_reg.entities.values()):
+            if (
+                entity_entry.config_entry_id == config_entry.entry_id
+                and entity_entry.domain == DEVICE_TRACKER_DOMAIN
+            ):
+                _LOGGER.debug("Removing entity %s", entity_entry.entity_id)
+
+                ent_reg.async_remove(entity_entry.entity_id)
```

Comment:
I think we only want to remove the entities that correspond to the devices with interface that is no longer tracked. We can do the same filtering as we do in `KeeneticRouter._update_devices`. When we have the mac address we can find the entity entry from the unique_id.

Reply:
The thing is that at this point in code it's possible to classify trackers into three categories:
1. entities that are currently connected to an interface that's not tracked anymore 
2. entities that are currently connected to an interface that's still tracked 
3. entities that are not connected currently

The last category actually contains devices that are *usually* connected (correspond) to either tracked interface or not tracked one.
The ideal goal would be to remove entities from the first category and a part of the third which correspond to interface that's not tracked anymore. With the possibilities stated above the best option is to remove the first and third categories, so that disconnected devices that are *usually* connected to still tracked interfaces will reappear when they connect next time.

Current implementation just drops every device so that the second category devices will reappear instantly on the integration reload. I see this technically the same effect as a doable solution above with the advantage of not having an extra async call to the router.

Reply:
We don't need to fetch more data from the router. We know the existing devices and their mac address. We just need to filter the devices with the interface that is no longer tracked and then remove the corresponding devices, using the mac address as id.